### PR TITLE
Enable Workgroup Reordering Based on Translation Info Config Entries

### PR DESCRIPTION
**Pull request by bangtianliu**
This pull request introduces the ability to reorder workgroups based on translation information specified in the configuration entries.

For instance, setting reorder_workgroups = "transpose" will enable the "transpose" method for workgroup reordering.

Notably, this pull request addresses a scenario that is somewhat the reverse of ISSUE #17340.

The "transpose" reordering method has been found to be faster than the natural order for convolution kernels.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp [hunk 1]
```diff
@@ -86,8 +86,26 @@ getPipelineOptions(FunctionOpInterface funcOp,
...
+      Attribute reorderGroupOption =
+          config.get(LLVMGPUAttrNames::kReorderWorkgroups);
+      assert(mlir::isa<mlir::StringAttr>(reorderGroupOption) &&
+             "reorder strategy should be a StringAttr");
+      StringRef reorderStr =
+          llvm::cast<StringAttr>(reorderGroupOption).getValue();
+      if (reorderStr == "transpose" || reorderStr == "Transpose") {
+        pipelineOptions.reorderOption =
+            LLVMGPUPipelineOptions::reorderWorkGroupOption::Transpose;
+      } else if (reorderStr == "swizzle" || reorderStr == "Swizzle") {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp]: I don't think we need to support both variants -- lowercase is fine.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp [hunk 2]
```diff
@@ -86,8 +86,26 @@ getPipelineOptions(FunctionOpInterface funcOp,
...
+        pipelineOptions.reorderOption =
+            LLVMGPUPipelineOptions::reorderWorkGroupOption::Swizzle;
+      } else {
+        pipelineOptions.reorderOption =
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp]: Here, should we check that the string is `none`? This is to diagnose cases when the value is misspelt.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 3]
```diff
@@ -77,7 +77,8 @@ static llvm::cl::opt<bool> clLLVMGPUEnablePrefetch(
 
 llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
                               const LLVMGPUPipelineOptions &options) {
-  return os << "{" << "enableReduceSharedMemoryBankConflicts = "
+  return os << "{"
+            << "enableReduceSharedMemoryBankConflicts = "
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: Why is this changed?

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.h [hunk 4]
```diff
@@ -25,15 +25,19 @@ namespace mlir::iree_compiler {
...
 /// function granularity.
 namespace LLVMGPUAttrNames {
-inline constexpr StringLiteral kNoReorderWorkgroups = "no_reorder_workgroups";
+inline constexpr StringLiteral kReorderWorkgroups = "reorder_workgroups";
 inline constexpr StringLiteral kNoReduceSharedMemoryBankConflicts =
     "no_reduce_shared_memory_bank_conflicts";
 } //  namespace LLVMGPUAttrNames
 
 struct LLVMGPUPipelineOptions {
+  enum reorderWorkGroupOption { None, Transpose, Swizzle };
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.h]:
```suggestion
  enum ReorderWorkgroupsOption { None, Transpose, Swizzle };
```

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.h [hunk 5]
```diff
@@ -25,15 +25,19 @@ namespace mlir::iree_compiler {
...
+
   bool enableReduceSharedMemoryBankConflicts = true;
-  bool enableReorderWorkgroups = true;
+  bool enableReorderWorkgroups = false;
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.h]: We don't need this anymore

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 6]
```diff
@@ -322,10 +323,24 @@ void addGPUMatmulSimtPassPipeline(OpPassManager &funcPassManager,
...
+        LLVMGPUPipelineOptions::reorderWorkGroupOption::Transpose)
+      reorderStrategy = ReorderWorkgrupsStrategy::Transpose;
+    else if (options.reorderOption ==
+             LLVMGPUPipelineOptions::reorderWorkGroupOption::Swizzle)
+      reorderStrategy = ReorderWorkgrupsStrategy::Swizzle;
+    else
+      reorderStrategy = ReorderWorkgrupsStrategy::None;
+    funcPassManager.addPass(createReorderWorkgroups(
+        reorderStrategy, clReorderWorkgroupsLogSwizzleTile,
+        canReorderWorkgroups));
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: This should probably go to a helper function

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 7]
```diff
@@ -77,7 +77,8 @@ static llvm::cl::opt<bool> clLLVMGPUEnablePrefetch(
 
 llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
                               const LLVMGPUPipelineOptions &options) {
-  return os << "{" << "enableReduceSharedMemoryBankConflicts = "
+  return os << "{"
+            << "enableReduceSharedMemoryBankConflicts = "
             << options.enableReduceSharedMemoryBankConflicts
             << ", enableReorderWorkgroups = " << options.enableReorderWorkgroups
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: The enum value is missing, we should also print it.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 8]
```diff
@@ -322,10 +323,24 @@ void addGPUMatmulSimtPassPipeline(OpPassManager &funcPassManager,
   }
 
   if (options.enableReorderWorkgroups) {
+    ReorderWorkgrupsStrategy reorderStrategy;
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: Can we use the same enum type and assign it directly?

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.h (hunk 4 above)
Comment (kuhar) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.h]: We should be careful so that we can distinguish the case when this disables workgroup reordering (when enabled globally) and when it's not set. I think we can use `std::optional<ReorderWorkgroupsStrategy>`

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/config_user_vector_distribute.mlir [hunk 9]
```diff
@@ -71,7 +71,7 @@ hal.executable public @main_0_dispatch_0 {
 
 // CHECK:       #[[$TRANSLATION:.+]] = #iree_codegen.translation_info<LLVMGPUVectorDistribute workgroup_size = [128, 2, 1] subgroup_size = 64
 // CHECK-SAME:    mma_schedule = #iree_gpu.mma_schedule<intrinsic = #iree_gpu.mma_layout<MFMA_F16_16x16x16_F32>,
-// CHECK-SAME:    no_reorder_workgroups
+// CHECK-SAME:    reorder_workgroups = "transpose"
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/config_user_vector_distribute.mlir]: I don't think this tests what it appears to since the attribute matches the global default. IMO we should maintain a test that disables globally-enables reordering (through the CLI flag) and then add a new test that enabled reordering when disabled globally.

**Review by kuhar — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp (hunk 3 above)
Comment (bangtianliu) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: adapting to changes in the data structure (LLVMGPUPipelineOptions).

Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: I think this just came from a different clang-format version. Same thing happened to me.

Comment (ScottTodd) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: Yep, clang-format wants this line unchanged: https://github.com/iree-org/iree/actions/runs/9472784149/job/26099808430?pr=17645#step:4:120

You can set up pre-commit to run formatting automatically or install a version close to what the CI uses: https://github.com/iree-org/iree/blob/c1e542d6370473244a8fa9178615cb8a6041b489/.pre-commit-config.yaml#L33-L41

Comment (bangtianliu) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: Thanks, I have set up pre-commit and it is very helpful.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp [hunk 10]
```diff
@@ -87,23 +87,20 @@ getPipelineOptions(FunctionOpInterface funcOp,
...
-        pipelineOptions.reorderOption =
-            LLVMGPUPipelineOptions::reorderWorkGroupOption::Swizzle;
+      if (reorderStr == "transpose") {
+        pipelineOptions.reorderStrategy = ReorderWorkgrupsStrategy::Transpose;
+      } else if (reorderStr == "swizzle") {
+        pipelineOptions.reorderStrategy = ReorderWorkgrupsStrategy::Swizzle;
       } else {
-        pipelineOptions.reorderOption =
-            LLVMGPUPipelineOptions::reorderWorkGroupOption::None;
+        assert(reorderStr == "none" && "Unhandled reorder option");
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp]: This should be an op error IMO -- the compiler shouldn't crash on invalid IR.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 11]
```diff
@@ -77,9 +77,28 @@ static llvm::cl::opt<bool> clLLVMGPUEnablePrefetch(
...
+                 ReorderWorkgrupsStrategy::None &&
+             "Unhandled reorder option");
+      reorderStr = "none";
+    }
+
+    return os << "{" << "enableReduceSharedMemoryBankConflicts = "
+              << options.enableReduceSharedMemoryBankConflicts
+              << ", ReorderWorkgroupsStrategy = " << reorderStr
+              << ", enableUkernels = " << options.enableUkernels << "}";
+  }
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: We should always print the strategy here. If the value is `std::nullopt` we can print something like `<not set>`

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.h [hunk 12]
```diff
@@ -25,15 +26,19 @@ namespace mlir::iree_compiler {
...
 /// function granularity.
 namespace LLVMGPUAttrNames {
-inline constexpr StringLiteral kNoReorderWorkgroups = "no_reorder_workgroups";
+inline constexpr StringLiteral kReorderWorkgroups = "reorder_workgroups";
 inline constexpr StringLiteral kNoReduceSharedMemoryBankConflicts =
     "no_reduce_shared_memory_bank_conflicts";
 } //  namespace LLVMGPUAttrNames
 
 struct LLVMGPUPipelineOptions {
+  //   enum reorderWorkGroupsOption { None, Transpose, Swizzle };
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.h]: Remove this

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.h [hunk 13]
```diff
@@ -25,15 +26,19 @@ namespace mlir::iree_compiler {
...
+
   bool enableReduceSharedMemoryBankConflicts = true;
-  bool enableReorderWorkgroups = true;
   bool enableUkernels = false;
+
+  //   reorderWorkGroupOption reorderOption = None;
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.h]: Remove this

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp [hunk 14]
```diff
@@ -86,8 +86,23 @@ getPipelineOptions(FunctionOpInterface funcOp,
   if (DictionaryAttr config = translationInfo.getConfiguration()) {
     if (config.contains(LLVMGPUAttrNames::kNoReduceSharedMemoryBankConflicts))
       pipelineOptions.enableReduceSharedMemoryBankConflicts = false;
-    if (config.contains(LLVMGPUAttrNames::kNoReorderWorkgroups))
-      pipelineOptions.enableReorderWorkgroups = false;
+    if (config.contains(LLVMGPUAttrNames::kReorderWorkgroups)) {
+      // Get the workgroups reorder config and enable the workgroup reordering
+      Attribute reorderGroupOption =
+          config.get(LLVMGPUAttrNames::kReorderWorkgroups);
+      assert(mlir::isa<mlir::StringAttr>(reorderGroupOption) &&
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp]:
```suggestion
      assert(isa<StringAttr>(reorderGroupOption) &&
```

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp [hunk 15]
```diff
@@ -86,8 +86,23 @@ getPipelineOptions(FunctionOpInterface funcOp,
...
+             "reorder strategy should be a StringAttr");
+      StringRef reorderStr =
+          llvm::cast<StringAttr>(reorderGroupOption).getValue();
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp]:
```suggestion
         cast<StringAttr>(reorderGroupOption).getValue();
```

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 16]
```diff
@@ -77,9 +77,28 @@ static llvm::cl::opt<bool> clLLVMGPUEnablePrefetch(
 
 llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
                               const LLVMGPUPipelineOptions &options) {
+  if (options.reorderStrategy) {
+    StringRef reorderStr;
+    if (options.reorderStrategy.value() == ReorderWorkgrupsStrategy::Transpose)
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: nit: you can compare directly with `==`
```suggestion
    if (options.reorderStrategy == ReorderWorkgrupsStrategy::Transpose)
```

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 17]
```diff
@@ -77,9 +77,28 @@ static llvm::cl::opt<bool> clLLVMGPUEnablePrefetch(
...
+      reorderStr = "transpose";
+    else if (options.reorderStrategy.value() ==
+             ReorderWorkgrupsStrategy::Swizzle)
+      reorderStr = "swizzle";
+    else {
+      assert(options.reorderStrategy.value() ==
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: nit: If one of the if-else branches has braces around it, the other ones should have them too

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/config_user_vector_distribute.mlir [hunk 18]
```diff
@@ -2,11 +2,20 @@
 // RUN:   --iree-codegen-reorder-workgroups-strategy=transpose \
 // RUN:   --pass-pipeline="builtin.module(hal.executable(hal.executable.variant(builtin.module(iree-llvmgpu-select-lowering-strategy, func.func(iree-llvmgpu-lower-executable-target)))))" %s | FileCheck %s
 
+// A new test that disables the global CLI flag (--iree-codegen-reorder-workgroups-strategy) and checks that applying reorder_workgroups = "transpose" enables workgroup reordering.
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/config_user_vector_distribute.mlir]:
```suggestion
// Check that applying `reorder_workgroups = "transpose"` enables workgroup reordering.
```

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/config_user_vector_distribute.mlir [hunk 19]
```diff
@@ -2,11 +2,20 @@
...
+
+// RUN: iree-opt --split-input-file --iree-gpu-test-target=gfx942 --iree-codegen-llvmgpu-use-vector-distribution \
+// RUN:   --pass-pipeline="builtin.module(hal.executable(hal.executable.variant(builtin.module(iree-llvmgpu-select-lowering-strategy, func.func(iree-llvmgpu-lower-executable-target)))))" %s | FileCheck %s --check-prefix=RWP
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/config_user_vector_distribute.mlir]: What does RWP stand for?

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/config_user_vector_distribute.mlir [hunk 20]
```diff
@@ -90,14 +110,24 @@ hal.executable public @main_0_dispatch_0 {
...
+      // RWP:         memref.alloc() : memref<128x36xf16, #gpu.address_space<workgroup>>
+      // RWP-DAG:     %[[WG_Y:.+]] = hal.interface.workgroup.id[1] : index
+      // RWP-DAG:     %[[WG_X:.+]] = hal.interface.workgroup.id[0] : index
+      // RWP-DAG:     arith.muli %[[WG_Y]], %{{.+}} : index
+      // RWP-DAG:     arith.addi %{{.+}}, %[[WG_X]] : index
+      // RWP:         scf.for  
       func.func @main_0_dispatch_0_matmul_transpose_b_2048x10240x1280_f16xf16xf32()
         attributes {translation_info = #iree_codegen.translation_info<LLVMGPUVectorDistribute workgroup_size = [128, 2, 1] subgroup_size = 64, {
           mma_schedule = #iree_gpu.mma_schedule<intrinsic = #iree_gpu.mma_layout<MFMA_F16_16x16x16_F32>, subgroup_m_count = 2, subgroup_n_count = 2>,
-          no_reorder_workgroups  // Disable the 'reorderWorkgroups' pass.
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/config_user_vector_distribute.mlir:100]: We should maintain a test where global reordering is disabled through this attribute

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/config_user_vector_distribute.mlir (hunk 19 above)
Comment (bangtianliu) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/config_user_vector_distribute.mlir]: Reorder-WorkgrouP

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp (hunk 15 above)
Comment (kuhar) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp]: not resolved

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp [hunk 21]
```diff
@@ -86,8 +86,25 @@ getPipelineOptions(FunctionOpInterface funcOp,
...
+      if (reorderStr == "transpose") {
+        pipelineOptions.reorderStrategy = ReorderWorkgrupsStrategy::Transpose;
+      } else if (reorderStr == "swizzle") {
+        pipelineOptions.reorderStrategy = ReorderWorkgrupsStrategy::Swizzle;
+      } else {
+        if (reorderStr != "none")
+          funcOp.emitOpError("Unhandled reorder option");
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp]: I don't remember the exact syntax, but this error can be made more helpful and precise:
```suggestion
          funcOp.emitOpError() << "Unknown " << LLVMGPUAttrNames::kReorderWorkgroups << "value: " << reorderGroupOpton;
```

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp [hunk 22]
```diff
@@ -86,8 +86,25 @@ getPipelineOptions(FunctionOpInterface funcOp,
   if (DictionaryAttr config = translationInfo.getConfiguration()) {
     if (config.contains(LLVMGPUAttrNames::kNoReduceSharedMemoryBankConflicts))
       pipelineOptions.enableReduceSharedMemoryBankConflicts = false;
-    if (config.contains(LLVMGPUAttrNames::kNoReorderWorkgroups))
-      pipelineOptions.enableReorderWorkgroups = false;
+    if (config.contains(LLVMGPUAttrNames::kReorderWorkgroups)) {
+      // Get the workgroups reorder config and enable the workgroup reordering
+      Attribute reorderGroupOption =
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp]:
```suggestion
      Attribute reorderWorkgroupOption =
```

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp [hunk 23]
```diff
@@ -86,8 +86,25 @@ getPipelineOptions(FunctionOpInterface funcOp,
   if (DictionaryAttr config = translationInfo.getConfiguration()) {
     if (config.contains(LLVMGPUAttrNames::kNoReduceSharedMemoryBankConflicts))
       pipelineOptions.enableReduceSharedMemoryBankConflicts = false;
-    if (config.contains(LLVMGPUAttrNames::kNoReorderWorkgroups))
-      pipelineOptions.enableReorderWorkgroups = false;
+    if (config.contains(LLVMGPUAttrNames::kReorderWorkgroups)) {
+      // Get the workgroups reorder config and enable the workgroup reordering
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp]:
```suggestion
      // Get the workgroups reorder config and enable the workgroup reordering.
```

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp [hunk 24]
```diff
@@ -86,8 +86,25 @@ getPipelineOptions(FunctionOpInterface funcOp,
...
+      Attribute reorderGroupOption =
+          config.get(LLVMGPUAttrNames::kReorderWorkgroups);
+      assert(isa<StringAttr>(reorderGroupOption) &&
+             "reorder strategy should be a StringAttr");
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerExecutableTarget.cpp]: Same here, this should be an error instead an assertion.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 25]
```diff
@@ -77,9 +77,22 @@ static llvm::cl::opt<bool> clLLVMGPUEnablePrefetch(
 
 llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
                               const LLVMGPUPipelineOptions &options) {
+  StringRef reorderStr = "< not set >";
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: uber nit:
```suggestion
  StringRef reorderStr = "<not set>";
```

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 26]
```diff
@@ -77,9 +77,22 @@ static llvm::cl::opt<bool> clLLVMGPUEnablePrefetch(
...
+      assert(options.reorderStrategy == ReorderWorkgrupsStrategy::None &&
+             "Unhandled reorder option");
+      reorderStr = "none";
+    }
+  }
+
   return os << "{" << "enableReduceSharedMemoryBankConflicts = "
             << options.enableReduceSharedMemoryBankConflicts
-            << ", enableReorderWorkgroups = " << options.enableReorderWorkgroups
+            << ", ReorderWorkgroupsStrategy = " << reorderStr
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]:
```suggestion
            << ", reorderWorkgroupsStrategy = " << reorderStr
```

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 27]
```diff
@@ -321,11 +334,16 @@ void addGPUMatmulSimtPassPipeline(OpPassManager &funcPassManager,
     funcPassManager.addPass(createGPUReduceBankConflictsPass());
   }
 
-  if (options.enableReorderWorkgroups) {
+  if (options.reorderStrategy) {
+    funcPassManager.addPass(createReorderWorkgroups(
+        options.reorderStrategy.value(), clReorderWorkgroupsLogSwizzleTile,
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]:
```suggestion
        *options.reorderStrategy, clReorderWorkgroupsLogSwizzleTile,
```

Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: We could have a helper function `getWokrgroupReoderingStrategy` that takes `options` and returns either the workgroup reordering strategy if set, or the global flag. This will reduce branching in this code and the code below.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/config_user_vector_distribute.mlir [hunk 28]
```diff
@@ -127,3 +157,64 @@ hal.executable public @main_0_dispatch_0 {
...
+          %10 = arith.mulf %8, %9 : f32
+          %11 = arith.addf %out, %10 : f32
+          linalg.yield %11 : f32
+        } -> tensor<2048x10240xf32>
+        flow.dispatch.tensor.store %7, %2, offsets = [0, 0], sizes = [2048, 10240], strides = [1, 1] : tensor<2048x10240xf32> -> !flow.dispatch.tensor<writeonly:tensor<2048x10240xf32>>
+        return
+      }
+    }
+  }
+}
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/config_user_vector_distribute.mlir]: missing newline

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 29]
```diff
@@ -183,6 +183,13 @@ static LogicalResult canReorderWorkgroups(FunctionOpInterface funcOp) {
   return success(workgroupCounts.size() >= 2);
 }
 
+static ReorderWorkgrupsStrategy getWorkgroupsReoderStrategy(
+    const std::optional<ReorderWorkgrupsStrategy> &option) {
+  if (option)
+    return *option;
+  return clReorderWorkgroupsStrategy;
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]:
```suggestion
  return option.value_or(clReorderWorkgroupsStrategy);
```

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 30]
```diff
@@ -183,6 +183,13 @@ static LogicalResult canReorderWorkgroups(FunctionOpInterface funcOp) {
   return success(workgroupCounts.size() >= 2);
 }
 
+static ReorderWorkgrupsStrategy getWorkgroupsReoderStrategy(
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]:
```suggestion
// Reconciles workgroup reordering strategy based on the pipeline `option` and the CLI flag.
static ReorderWorkgrupsStrategy getWorkgroupsReoderStrategy(
```

**Review by kuhar — COMMENTED** [review]
LGTM % nits. I will give it a try before approving.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/config_user_vector_distribute.mlir [hunk 31]
```diff
@@ -87,17 +107,88 @@ hal.executable public @main_0_dispatch_0 {
...
+      // OPT-OUT:         memref.alloc() : memref<128x36xf16, #gpu.address_space<workgroup>>
+      // OPT-OUT:         memref.alloc() : memref<128x36xf16, #gpu.address_space<workgroup>>
+      // OPT-OUT-DAG:     hal.interface.workgroup.id[1] : index
+      // OPT-OUT-DAG:     hal.interface.workgroup.id[0] : index
+      // OPT-OUT-NEXT:    scf.for
       func.func @main_0_dispatch_0_matmul_transpose_b_2048x10240x1280_f16xf16xf32()
         attributes {translation_info = #iree_codegen.translation_info<LLVMGPUVectorDistribute workgroup_size = [128, 2, 1] subgroup_size = 64, {
           mma_schedule = #iree_gpu.mma_schedule<intrinsic = #iree_gpu.mma_layout<MFMA_F16_16x16x16_F32>, subgroup_m_count = 2, subgroup_n_count = 2>,
-          no_reorder_workgroups  // Disable the 'reorderWorkgroups' pass.
+          reorder_workgroups = "none"  // disable the 'reorderWorkgroups' pass.
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/config_user_vector_distribute.mlir]:
```suggestion
          reorder_workgroups = "none"  // Disable the 'reorderWorkgroups' pass.
```

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.h [hunk 32]
```diff
@@ -25,15 +26,16 @@ namespace mlir::iree_compiler {
...
 inline constexpr StringLiteral kNoReduceSharedMemoryBankConflicts =
     "no_reduce_shared_memory_bank_conflicts";
 } //  namespace LLVMGPUAttrNames
 
 struct LLVMGPUPipelineOptions {
   bool enableReduceSharedMemoryBankConflicts = true;
-  bool enableReorderWorkgroups = true;
   bool enableUkernels = false;
+
+  std::optional<ReorderWorkgroupsStrategy> reorderStrategy;
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.h]:
```suggestion
  bool enableUkernels = false;
  std::optional<ReorderWorkgroupsStrategy> reorderStrategy;
```

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 33]
```diff
@@ -170,6 +183,15 @@ static LogicalResult canReorderWorkgroups(FunctionOpInterface funcOp) {
   return success(workgroupCounts.size() >= 2);
 }
 
+// Reconciles workgroup reordering strategy based on the pipeline `option` and
+// the CLI flag.
+static ReorderWorkgroupsStrategy getWorkgroupsReoderStrategy(
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]:
```suggestion
static ReorderWorkgroupsStrategy getReorderWorkgroupsStrategy(
```

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp [hunk 34]
```diff
@@ -170,6 +183,15 @@ static LogicalResult canReorderWorkgroups(FunctionOpInterface funcOp) {
...
+    const std::optional<ReorderWorkgroupsStrategy> &option) {
+  if (option)
+    return *option;
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp]: Delete this
```suggestion
```

**Review by kuhar — APPROVED** [review]
LGTM. I can see this improves performance on SDXL convs. Thanks for all the fixes.

**Comment by bangtianliu** [comment]
> LGTM. I can see this improves performance on SDXL convs. Thanks for all the fixes.

Thanks for your comments and help, appreciate it very much!

**Comment by saienduri** [comment]
Hi @bangtianliu, can you also update the spec in this PR: https://github.com/iree-org/iree/blob/main/build_tools/pkgci/external_test_suite/attention_and_matmul_spec.mlir? We need to keep it in sync with sdxl-scripts repo so we pick up the same perf boosts here as well
